docs/source/sparse.rst

```diff
@@ -1,146 +1,382 @@
+           values=tensor(  [3, 4]),
+           size=(3,), nnz=2, layout=torch.sparse_coo)
+    >>> s.to_dense()
+    tensor([0, 7, 0])
 
-SparseTensor has the following invariants:
-  1. sparse_dim + dense_dim = len(SparseTensor.shape)
-  2. SparseTensor._indices().shape = (sparse_dim, nnz)
-  3. SparseTensor._values().shape = (nnz, SparseTensor.shape[sparse_dim:])
+Uncoalesced tensors permit us to implement certain operators more
+efficiently.  For the most part, you shouldn't have to care whether or
+not a sparse tensor is coalesced or not, as most operations will work
+identically given a coalesced or uncoalesced sparse tensor.  However,
+there are two cases in which you may need to care.
 
-Since SparseTensor._indices() is always a 2D tensor, the smallest sparse_dim = 1.
-Therefore, representation of a SparseTensor of sparse_dim = 0 is simply a dense tensor.
+First, if you repeatedly perform an operation that can produce
+duplicate entries (e.g., :func:`torch.Tensor.add`), you should
+occasionally coalesce your sparse tensors to prevent them from growing
+too large:
+
+    >>> s.add_(torch.sparse_coo_tensor([[2, 2]], [5, 6]))
+    tensor(indices=tensor([[1, 1, 2, 2]]),
+           values=tensor(  [3, 4, 5, 6]),
+           size=(3,), nnz=4, layout=torch.sparse_coo)
+    >>> s.coalesce()
+    tensor(indices=tensor([[1,  2]]),
+           values=tensor(  [7, 11]),
+           size=(3,), nnz=2, layout=torch.sparse_coo)
+
+.. note::
+
+   A coalesced tensor is a sparse COO tensor with unique indices in
+   lexicographical ordering.
+
+Second, some operators can be implemented on uncoalesced sparse
+tensors. The corresponding implementations would use
+:func:`torch.Tensor._values()` and :func:`torch.Tensor._indices()` to
+retrieve the uncoalesced data of a tensor. One must take into an
+account the additive nature of uncoalesced data: the values of the
+same indices are the terms of a sum that evaluation gives the value of
+the corresponding tensor element. For example, the scalar
+multiplication on an uncoalesced sparse tensor could be implemented by
+multiplying all the uncoalesced values with the scalar because ``c *
+(a + b) == c * a + c * b`` holds. However, any nonlinear operation,
+say, a square root, cannot be implemented on applying the operation to
+uncoalesced data because ``sqrt(a + b) == sqrt(a) + sqrt(b)`` does not
+hold, for instance.
+
+Linear Algebra operations
++++++++++++++++++++++++++
+
+The following table summarizes supported Linear Algebra operations on
+sparse matrices with different storage layouts. Here ``T[layout]``
+denotes a tensor with a give layout. Similarly, ``M[layout]`` denotes
+a matrix (2-D Torch tensor), and ``V[layout]`` denotes a vector (1-D
+Torch tensor). In addition, ``f`` denotes a scalar (float or 0-D Torch
+tensor), ``*`` is element-wise multiplication, ``@`` is matrix
+multiplication.
+
+.. csv-table::
+   :header: "Torch operation", "Sparse grad?", "Layout signature"
```

Comment:
Speaking of autograd, is there a section in this document on sparse autograd?

Reply:
No, and I think there is no need for a separate "Sparse Autograd" section. At the user level, Autograd is not a storage format dependent concept.

Here, "Sparse grad?" describes if the Autograd support has been implemented for given functions that supports sparse tensors as inputs.

Supporting Autograd for functions on sparse inputs can be a technically complex topic but at the user-facing documentation level, the users only need to know if a particular function supports Autograd or not.